oracle/provider/uniswap.go

```diff
@@ -0,0 +1,338 @@
+			Token1             Token   `graphql:"token1"`
+			Token0Price        string  `graphql:"token0Price"`
+			Token1Price        string  `graphql:"token1Price"`
+			VolumeUSDTracked   string  `graphql:"volumeUSDTracked"`
+			VolumeUSDUntracked string  `graphql:"volumeUSDUntracked"`
+		} `graphql:"poolHourDatas(first: $first,after: $after, orderBy: periodStartUnix, orderDirection: desc, where: {poolID_in: $poolIDS, periodStartUnix_gte:$start,periodStartUnix_lte:$stop})"` //nolint:lll
+	}
+
+	// UniswapProvider defines an Oracle provider implemented to consume data from Uniswap graphql
+	UniswapProvider struct {
+		baseURL        string
+		client         *gql.Client
+		denomToAddress map[string]string
+	}
+)
+
+func NewUniswapProvider(endpoint Endpoint, currencyPairs ...types.CurrencyPair) *UniswapProvider {
+	// create pair name to address map
+	denomToAddress := make(map[string]string)
+	for _, pair := range currencyPairs {
+		// graph supports all lower case id's
+		// currently supports only 1 fee tier pool per currency pair
+		address := strings.ToLower(pair.Address)
+		denomToAddress[pair.String()] = address
+	}
+
+	// default provider to eth uniswap
+	if endpoint.Name == ProviderEthUniswap {
+		return &UniswapProvider{
+			baseURL:        endpoint.Rest,
+			client:         gql.NewClient(endpoint.Rest, nil),
+			denomToAddress: denomToAddress,
+		}
+	}
+
+	return &UniswapProvider{
+		baseURL:        UniswapURL,
+		client:         gql.NewClient(UniswapURL, nil),
+		denomToAddress: denomToAddress,
+	}
+}
+
+func (p *UniswapProvider) StartConnections() {
+	// no-op Uniswap v1 does not use websockets
+}
+
+// SubscribeCurrencyPairs performs a no-op since Uniswap does not use websockets
+func (p UniswapProvider) SubscribeCurrencyPairs(...types.CurrencyPair) {}
+
+func (p UniswapProvider) GetTickerPrices(pairs ...types.CurrencyPair) (map[string]types.TickerPrice, error) {
+	// create pool ids
+	poolIDS, err := p.collectPoolIDS(pairs...)
+	if err != nil {
+		return nil, err
+	}
+
+	idMap := map[string]interface{}{
+		"poolIDS": poolIDS,
+		"start":   time.Now().Unix() - 86400,
+		"stop":    time.Now().Unix(),
+	}
+
+	baseDenomIdx, quoteDenomIdx := p.returnBaseAndQuoteMapping(pairs...)
+	tickerPrices := make(map[string]types.TickerPrice, len(pairs))
+	latestTimestamp := make(map[string]float64)
+
+	var lastID string
+	var firstID string
+	var poolsHourDatas PoolHourDataQuery
+	for {
+		// limit by graph
+		idMap["first"] = 1000
+		idMap["after"] = lastID
+
+		// query volume from day data
+		var poolsHourData PoolHourDataQuery
+		err := p.client.Query(context.Background(), &poolsHourData, idMap)
+
+		if err != nil {
+			return nil, err
+		}
+
+		// check if no new id or repeated id
+		if len(poolsHourData.PoolHourDatas) == 0 || firstID == poolsHourData.PoolHourDatas[0].ID {
+			break
+		}
+
+		firstID = poolsHourData.PoolHourDatas[0].ID
+		lastID = poolsHourData.PoolHourDatas[len(poolsHourData.PoolHourDatas)-1].ID
```

Comment:
This query seems to happen in the same thread as the main price-feeder loop that gathers all the prices, calculates tvwaps, and sends them to broadcast. My concern is the price feeder will get slower for each uniswap pair we add. Or what if the connection to uniswap hangs, then this price-feeder will stop submitting prices until the timeout is reached. Do you think it would be worthwhile to move polling uniswap to its own go routine. It could be helpful down the road when we start listing hundreds of assets to keep that main loop running efficiently.